runtime/src/iree/hal/drivers/hip/pending_queue_actions.c

```diff
@@ -541,35 +552,36 @@ static void iree_hal_hip_execution_device_signal_host_callback(
     void* user_data) {
   IREE_TRACE_ZONE_BEGIN(z0);
   iree_hal_hip_queue_action_t* action = (iree_hal_hip_queue_action_t*)user_data;
-  IREE_ASSERT_EQ(action->kind, IREE_HAL_HIP_QUEUE_ACTION_TYPE_EXECUTION);
-  IREE_ASSERT_EQ(action->state, IREE_HAL_HIP_QUEUE_ACTION_STATE_ALIVE);
   iree_hal_hip_pending_queue_actions_t* actions = action->owning_actions;
 
+  iree_slim_mutex_lock(&actions->action_mutex);
+  IREE_ASSERT_EQ(action->kind, IREE_HAL_HIP_QUEUE_ACTION_TYPE_EXECUTION);
+  IREE_ASSERT_EQ(action->state, IREE_HAL_HIP_QUEUE_ACTION_STATE_ALIVE);
   // Flip the action state to zombie and enqueue it again so that we can let
   // the worker thread clean it up. Note that this is necessary because cleanup
   // may involve GPU API calls like buffer releasing or unregistering, so we can
   // not inline it here.
   action->state = IREE_HAL_HIP_QUEUE_ACTION_STATE_ZOMBIE;
-  iree_slim_mutex_lock(&actions->action_mutex);
   iree_hal_hip_queue_action_list_push_back(&actions->action_list, action);
   iree_slim_mutex_unlock(&actions->action_mutex);
 
-  // Notify the worker thread again that we have the cleanup action enqueued.
-  // Only overwrite the idle waiting state, which has lower priority.
-  iree_hal_hip_worker_state_t prev_state =
-      IREE_HAL_HIP_WORKER_STATE_IDLE_WAITING;
-  iree_atomic_compare_exchange_strong_int32(
-      &actions->working_area.worker_state, /*expected=*/&prev_state,
-      /*desired=*/IREE_HAL_HIP_WORKER_STATE_WORKLOAD_PENDING,
-      /*order_succ=*/iree_memory_order_acq_rel,
-      /*order_fail=*/iree_memory_order_acquire);
-  iree_notification_post(&actions->working_area.state_notification,
-                         IREE_ALL_WAITERS);
+  iree_status_t status;
+  if (action->signal_semaphore_list.count) {
+    // Advance semaphore timelines by calling into the host signaling function.
+    // This will internally try to release more workload to the GPU.
+    status = iree_hal_semaphore_list_signal(action->signal_semaphore_list);
+  } else {
+    // If there is no semaphores to signal we still need to trigger execution
+    // of the recently zombified action so it can be cleaned up.
+    status = iree_hal_hip_pending_queue_actions_issue(actions);
+  }
 
-  // Advance semaphore timelines by calling into the host signaling function.
-  // This will internally try to release more workload to the GPU.
-  IREE_IGNORE_ERROR(
-      iree_hal_semaphore_list_signal(action->signal_semaphore_list));
+  if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+    IREE_ASSERT(false &&
+                "Can't signal semaphores and/or issue pending actions");
```

Comment:
Nit: no captialization of the first character: "cannot signal .."

Reply:
I removed it.